path_role.go

```diff
@@ -0,0 +1,431 @@
+// does not acquire the read lock before reading the role from the storage. If
+// locking is desired, use lockedOCIRole instead.
+// This method also does NOT check to see if a role upgrade is required. It is
+// the responsibility of the caller to check if a role upgrade is required and,
+// if so, to upgrade the role
+func (b *backend) nonLockedOCIRole(ctx context.Context, s logical.Storage, roleName string) (*OCIRoleEntry, error) {
+	if roleName == "" {
+		return nil, fmt.Errorf("missing role name")
+	}
+
+	entry, err := s.Get(ctx, "role/"+strings.ToLower(roleName))
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+
+	var result OCIRoleEntry
+	if err := entry.DecodeJSON(&result); err != nil {
+		return nil, err
+	}
+
+	return &result, nil
+}
+
+func (b *backend) pathRoleDelete(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role"), nil
+	}
+
+	b.roleMutex.Lock()
+	defer b.roleMutex.Unlock()
+
+	return nil, req.Storage.Delete(ctx, "role/"+strings.ToLower(roleName))
+}
+
+func (b *backend) pathRoleList(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	b.roleMutex.RLock()
+	defer b.roleMutex.RUnlock()
+
+	roles, err := req.Storage.List(ctx, "role/")
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(roles), nil
+}
+
+func (b *backend) pathRoleRead(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleEntry, err := b.lockedOCIRole(ctx, req.Storage, strings.ToLower(data.Get("role").(string)))
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return nil, nil
+	}
+
+	return &logical.Response{
+		Data: roleEntry.ToResponseData(),
+	}, nil
+}
+
+// create a Role
+func (b *backend) pathRoleCreate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	roleName := strings.ToLower(data.Get("role").(string))
+	if roleName == "" {
+		return logical.ErrorResponse("missing role"), nil
+	}
+
+	if len(roleName) > MaxRoleNameLength {
+		return logical.ErrorResponse("role length exceeds the limit"), nil
+	}
+
+	validateRoleRegEx := regexp.MustCompile(`^[A-Za-z0-9]+$`).MatchString
+	if !validateRoleRegEx(roleName) {
+		return logical.ErrorResponse("role is invalid"), nil
+	}
+
+	b.roleMutex.Lock()
+	defer b.roleMutex.Unlock()
+
+	roleEntry, err := b.nonLockedOCIRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+
+	if roleEntry != nil {
+		return logical.ErrorResponse("The specified role already exists"), nil
+	}
+
+	roleEntry = &OCIRoleEntry{
+		Role:    roleName,
+		Version: 1,
+	}
+
+	if description, ok := data.GetOk("description"); ok {
+		roleEntry.Description = description.(string)
+	}
+
+	if add_ocid_list, ok := data.GetOk("add_ocid_list"); ok {
```

Comment:
Use camelCase for the variable. (The parameter is fine as snake_case.)

Reply:
Done